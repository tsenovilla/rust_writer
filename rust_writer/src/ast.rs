// SPDX-License-Identifier: GPL-3.0

pub mod finder;
pub mod implementors;
pub mod macros;
pub mod mutator;
