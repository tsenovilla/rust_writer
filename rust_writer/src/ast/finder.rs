// SPDX-License-Identifier: GPL-3.0

//! This module contains the [`Finder`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/struct.Finder.html)
//! struct, which is used to search for specific items in an AST in a very targeted way.
//! The `Finder` struct is completely generic, and thanks to the
//! [`ToFind`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/trait.ToFind.html) trait and the
//! [implementors](https://docs.rs/rust_writer/latest/rust_writer/ast/implementors/index.html) it
//! may be customized to assert if an element is contained inside an AST.
//!
//! # Example
//! ```rust
//! use test_builder::TestBuilder;
//! use rust_writer::ast::{
//!     finder::{Finder, ToFind},
//!     implementors::ItemToTrait
//! };
//! use syn::{
//!     visit::Visit,
//!     parse_quote,
//!     File
//! };
//!
//! TestBuilder::default()
//!     .with_trait_ast()
//!     .execute(|builder| {
//!         let ast = builder.get_ref_ast_file("trait.rs").expect("This exists; qed;");
//!
//!         // Define the ItemFinder implementor. This implementor specifies:
//!         // 1. The trait we want to search for, e.g., a trait named `MyTrait`.
//!         // 2. The specific item(s) we expect to find within that trait.
//!         let item_to_trait: ItemToTrait =
//!             ("MyTrait", parse_quote! { type Type1: From<String>; }).into();
//!
//!         // Create the Finder and load this particular implementor
//!         let mut finder = Finder::default().to_find(&item_to_trait);
//!
//!         // Search the AST for the expected item
//!         assert!(finder.find(&ast));
//!     });
//! ```

#[cfg(test)]
mod tests;

use std::fmt::Debug;
use syn::{visit::Visit, File};

/// A placeholder finder which does not perform any specific search.
/// This is used as the default for a [`Finder`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/struct.Finder.html).
#[derive(Debug)]
pub struct EmptyFinder;

/// The `Finder` struct is used to assert if a concrete element belongs to an AST.
/// The `Finder` struct is totally generic, so it can be thought of a game console: its purpose is
/// to determine if an element is contained inside an AST but it doesn't know how to do it until an
/// [implementor](https://docs.rs/rust_writer/latest/rust_writer/ast/implementors/index.html)
/// is loaded. This implementor will tell `Finder` how it should behave.
///
/// A new `Finder` can be easily created using `Finder::default()`, which returns a
/// `Finder<'_, EmptyFinder, 1>`. This `Finder` is useless up to this point, but thanks to the
/// [`ToFind`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/trait.ToFind.html)
/// trait, an implementor can be seamlessly loaded to the `Finder`, releasing its whole power.
///
/// Once configured, the [`find`](#method.find) method can be called to look inside the AST.
///
/// ```rust
/// use test_builder::TestBuilder;
/// use rust_writer::ast::{
///     finder::{Finder, ToFind},
///     implementors::ItemToTrait
/// };
/// use syn::{
///     visit::Visit,
///     parse_quote,
///     File
/// };
///
/// TestBuilder::default()
///     .with_trait_ast()
///     .execute(|builder| {
///         let ast = builder.get_ref_ast_file("trait.rs").expect("This exists; qed;");
///
///         // Define the ItemFinder implementor. This implementor specifies:
///         // 1. The trait we want to search for, e.g., a trait named `MyTrait`.
///         // 2. The specific item(s) we expect to find within that trait.
///         let item_to_trait: ItemToTrait =
///             ("MyTrait", parse_quote! { type Type1: From<String>; }).into();
///
///         // Create the Finder and load this particular implementor
///         let mut finder = Finder::default().to_find(&item_to_trait);
///
///         // Search the AST for the expected item
///         assert!(finder.find(&ast));
///     });
/// ```
#[derive(Debug, Clone)]
pub struct Finder<'a, T: Debug, const N: usize> {
	/// A `Finder` can act in different parts of the AST at the same time. This array keeps track
	/// of which of these searchs succeeded.  
	pub found: [bool; N],
	/// Placeholder to load an implementor.
	pub finder: &'a T,
}

/// This trait is used to create a new `Finder` variable.
/// It is typically implemented for `Finder<'_, EmptyFinder, 1>`, enabling the flow:
/// 1. Create a `Finder<'_, EmptyFinder, 1>` using `Finder::default()`.
/// 2. Configure it with the desired implementor using the `to_find` method.
pub trait ToFind<'a, T: Debug, const N: usize> {
	fn to_find(self, finder: &'a T) -> Finder<'a, T, N>;
}

impl Default for Finder<'_, EmptyFinder, 1> {
	fn default() -> Self {
		Self { found: [false], finder: &EmptyFinder }
	}
}

impl<'a, T, const N: usize> Finder<'a, T, N>
where
	T: Debug,
	Finder<'a, T, N>: Visit<'a>,
{
	/// Apply all the searches defined by the implementor.
	pub fn find(&mut self, ast: &'a File) -> bool {
		self.visit_file(ast);
		self.found.iter().all(|&x| x)
	}
}

impl<T, const N: usize> Finder<'_, T, N>
where
	T: Debug,
{
	/// Reset the `found` array to all `false` values.
	pub fn reset(&mut self) {
		self.found = [false; N];
	}
}
