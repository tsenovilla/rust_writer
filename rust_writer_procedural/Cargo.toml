[package]
name = "rust_writer_procedural"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["full", "extra-traits", "parsing"] }

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.103"
rust_writer = { workspace = true }
test_builder = { workspace = true }
