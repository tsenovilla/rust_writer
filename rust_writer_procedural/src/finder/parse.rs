// SPDX-License-Identifier: GPL-3.0

use crate::parse_attrs::{MacroAttr, MacroAttrs};
use syn::{Error, Fields, FieldsNamed, ItemStruct, Path, Result, Type, TypeArray};

pub(crate) struct FinderDef {
	pub(crate) crate_implementors: Vec<Path>,
	pub(crate) local_implementors: Vec<Path>,
	pub(crate) struct_: ItemStruct,
	pub(crate) already_expanded: bool,
	pub(crate) impl_from: bool,
}

pub(crate) struct ImplFinderDef {
	pub(crate) struct_: ItemStruct,
}

impl FinderDef {
	pub(crate) fn try_from(attrs: MacroAttrs, mut struct_: ItemStruct) -> Result<Self> {
		let (already_expanded, impl_from) = attrs.validate_struct(&mut struct_)?.parse();

		let mut crate_implementors: Vec<Path> = Vec::with_capacity(attrs.0.len());
		let mut local_implementors: Vec<Path> = Vec::with_capacity(attrs.0.len());
		attrs.0.into_iter().for_each(|macro_attr| match macro_attr {
			MacroAttr::CrateImplementor(path) => crate_implementors.push(path),
			MacroAttr::LocalImplementor(path) => local_implementors.push(path),
		});

		Ok(Self { crate_implementors, local_implementors, struct_, already_expanded, impl_from })
	}
}

impl ImplFinderDef {
	pub(crate) fn try_from(struct_: ItemStruct) -> Result<Self> {
		match &struct_.fields {
			Fields::Named(FieldsNamed { named, .. })
				if named.iter().any(|field| {
					field.ident.as_ref().expect("Named fields have ident; qed;") == "found" &&
						matches!(&field.ty, Type::Array(TypeArray { elem, .. })
							if matches!(&**elem, Type::Path(path) if path.path.is_ident("bool"))
						)
				}) =>
				Ok(Self { struct_ }),
			_ => Err(Error::new(
				struct_.ident.span(),
				"Expected a file named found being [bool;N] inside struct.",
			)),
		}
	}
}
