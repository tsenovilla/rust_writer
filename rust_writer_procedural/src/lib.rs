// SPDX-License-Identifier: GPL-3.0

mod finder;
pub(crate) mod helpers;
mod mutator;
pub(crate) mod parse;

use proc_macro::TokenStream;

/// # Description
///
/// The `#[mutator]` macro is used to define a new implementor which combines other implementors,
/// capable of mutating different parts of an AST with just one instruction, with the same level of
/// precision as if we had used each implementor separately.
///
/// The syntax to use the `#[mutator]` macro is pretty simple:
///
/// ```rust
/// use rust_writer::ast::{mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::visit_mut::VisitMut;
///
/// // The new implementor behaves as two `ItemToTrait` implementors.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>)]
/// struct NewImplementor;
/// ```
///
/// From the snippet above we can already assert that the `#[mutator]` macro needs that the traits
/// [`ToMutate`](https://docs.rs/rust_writer/latest/rust_writer/ast/mutator/trait.ToMutate.html)
/// and [`VisitMut`](https://docs.rs/syn/latest/syn/visit_mut/trait.VisitMut.html) need to be
/// brought into scope.
///
/// It's also possible to combine an implementor from the predefined set with a local implementor,
/// just by adding the keyword `local` before the local implementor.
///
/// ```rust
/// use rust_writer::ast::{mutator, local_mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::{visit_mut::VisitMut, File};
///
/// #[local_mutator]
/// #[derive(Debug, Clone)]
/// struct LocalImplementor<T: std::fmt::Debug + Clone>{
///   mutated: [bool; 4],
///   something: T
/// }
///
/// impl<T> VisitMut for LocalImplementor<T> where T: std::fmt::Debug + Clone{
///   fn visit_file_mut(&mut self, _file: &mut File){
///     // Dummy implementation
///     self.mutated = [true; 4];
///   }
/// }
///
/// // The new implementor behaves as two `ItemToTrait` implementors + our local implementor.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>, local = LocalImplementor<T: std::fmt::Debug + Clone>)]
/// struct NewImplementor;
/// ```
///
/// When the `#[mutator]` macro is called, the struct is expanded an includes a new field for each
/// implementor named as the implementor type in lowercase. If an implementor is included several
/// times in the same macro invocation, a suffix is added to the subsequent fields, such as
/// _1, _2, _3 and so on. For example, the snippet above struct will look like this after
/// the expansion:
///
/// ```ignore
/// struct NewImplementor<'a, T: std::fmt::Debug + Clone>{
///   itemtotrait: ItemToTrait<'a>,
///   itemtotrait_1: ItemToTrait<'a>,
///   localimplementor: LocalImplementor<T>
/// }
/// ```
///
/// Awesome, these snippets also show that if an implementor needs a generic, including it in the
/// macro invocation is enough. Particularly, the generics will be added to the `NewImplementor`
/// struct, so they'll be available inside it. It's also possible to place.
///
/// # The `#[impl_from]` attribute
///
/// # Compatibility with [`#[finder]`](https://docs.rs/rust_writer/latest/rust_writer/ast/attr.finder.html)
#[proc_macro_attribute]
pub fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	mutator::mutator(attrs, item)
}

#[proc_macro_attribute]
pub fn local_mutator(_: TokenStream, item: TokenStream) -> TokenStream {
	mutator::local_mutator(item)
}

#[proc_macro_attribute]
pub fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::finder(attrs, item)
}

#[proc_macro_attribute]
pub fn local_finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::local_finder(attrs, item)
}

#[doc(hidden)]
#[proc_macro_attribute]
pub fn already_expanded(_: TokenStream, item: TokenStream) -> TokenStream {
	item
}
