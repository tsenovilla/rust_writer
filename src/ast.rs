// SPDX-License-Identifier: GPL-3.0

pub mod finder;
pub mod implementors;
pub mod mutator;
