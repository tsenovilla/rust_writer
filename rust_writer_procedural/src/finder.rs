// SPDX-License-Identifier: GPL-3.0

mod expand;

use crate::parse::{MacroAttrs, MacroImplParsed, MacroParsed};
use proc_macro::TokenStream;
use syn::{parse_macro_input, ItemStruct};

pub(crate) fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	let attrs_list = parse_macro_input!(attrs as MacroAttrs);
	let struct_def = parse_macro_input!(item as ItemStruct);

	match MacroParsed::try_from(attrs_list, struct_def) {
		Ok(parsed) => expand::expand_finder(parsed).into(),
		Err(err) => err.to_compile_error().into(),
	}
}

pub(crate) fn impl_finder(item: TokenStream) -> TokenStream {
	let mut finished = item.clone();

	let struct_def = parse_macro_input!(item as ItemStruct);

	let generated: TokenStream = match MacroImplParsed::try_from(struct_def, "found") {
		Ok(parsed) => expand::expand_impl_finder(parsed).into(),
		Err(err) => err.to_compile_error().into(),
	};

	finished.extend(generated);
	finished
}
