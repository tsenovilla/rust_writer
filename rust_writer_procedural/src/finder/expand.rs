// SPDX-License-Identifier: GPL-3.0

#[cfg(test)]
mod tests;

use crate::parse::{MacroImplParsed, MacroParsed};
use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::{
	parse_quote, punctuated::Punctuated, token::Brace, Field, Fields, FieldsNamed, GenericParam,
	Ident, Index, Lifetime, LifetimeParam, Token, Type,
};

pub(crate) fn expand_finder(parsed: MacroParsed) -> TokenStream {
	let MacroParsed {
		crate_implementors_idents,
		local_implementors_idents,
		mut struct_,
		already_expanded,
		impl_from,
		one,
		implementors_count,
		crate_implementors_indexes,
		local_implementors_indexes,
		generics_idents,
		where_clause,
		new_struct_fields,
	} = parsed;

	let struct_vis = &struct_.vis;
	let struct_name = &struct_.ident;

	let finder_wrapper_name =
		Ident::new(&(struct_.ident.to_string() + "FinderWrapper"), Span::call_site());

	let finder_lifetime: Lifetime = parse_quote! {'finder};

	// Expand struct fields and return a reference to the inner fields
	let struct_fields: &Punctuated<Field, Token![,]> = match struct_.fields {
		Fields::Unit => {
			struct_.fields = Fields::Named(FieldsNamed {
				brace_token: Brace::default(),
				named: new_struct_fields.clone(),
			});
			&new_struct_fields
		},
		Fields::Named(ref mut fields) => {
			fields.named.extend(new_struct_fields);
			&fields.named
		},
		_ => unreachable!("Parser doesn't allow Unnamed fields; qed;"),
	};

	let mut impl_from_block = quote! {};

	if impl_from {
		let fields_names: Vec<&Ident> = struct_fields
			.iter()
			.map(|field| field.ident.as_ref().expect("Named field has ident; qed;"))
			.collect();

		let fields_types: Punctuated<&Type, Token![,]> =
			struct_fields.iter().map(|field| &field.ty).collect();
		let tuple_indexes: Vec<Index> =
			struct_fields.iter().enumerate().map(|(index, _)| Index::from(index)).collect();

		impl_from_block = quote! {
			impl<#generics_idents> From<(#fields_types)> for #struct_name<#generics_idents> #where_clause{
				fn from(tuple: (#fields_types)) -> Self{
					Self{ #(#fields_names: tuple.#tuple_indexes),* }
				}
			}
		};

		struct_.attrs.push(parse_quote!(#[rust_writer_procedural::already_impl_from]));
	}

	let finder_wrapper = quote! {
		#[derive(Debug, Clone)]
		#struct_vis struct #finder_wrapper_name<#finder_lifetime, #generics_idents>(
			#struct_vis rust_writer::ast::finder::Finder<
				#finder_lifetime,
				#struct_name<#generics_idents>,
				#implementors_count
			>
		) #where_clause;

		impl<#finder_lifetime, #generics_idents> From<
			rust_writer::ast::finder::Finder<#finder_lifetime, #struct_name<#generics_idents>, #implementors_count>
		> for #finder_wrapper_name<#finder_lifetime, #generics_idents> #where_clause{
			#struct_vis fn from(input: rust_writer::ast::finder::Finder<
				#finder_lifetime,
				#struct_name<#generics_idents>,
				#implementors_count
			>) -> Self {
				Self(input)
			}
		}
	};

	let impl_to_find = quote! {
		impl<#finder_lifetime, #generics_idents>
		rust_writer::ast::finder::ToFind<#finder_lifetime, #struct_name<#generics_idents>, #implementors_count>
		for rust_writer::ast::finder::Finder<'_, rust_writer::ast::finder::EmptyFinder, #one>
		#where_clause
		{
			fn to_find(self, finder: &#finder_lifetime #struct_name<#generics_idents>)
			->
			rust_writer::ast::finder::Finder<#finder_lifetime, #struct_name<#generics_idents>,#implementors_count> {
				rust_writer::ast::finder::Finder {
					found: [false; #implementors_count],
					finder
				}
			}
		}
	};

	let impl_visit = quote! {
		impl<#finder_lifetime, #generics_idents>
		syn::visit::Visit<#finder_lifetime>
		for #finder_wrapper_name<#finder_lifetime, #generics_idents>
		#where_clause
		{
			fn visit_file(&mut self, file: &#finder_lifetime syn::File){
				#(
					let mut finder = rust_writer::ast::finder::Finder::default()
						.to_find(&self.0.finder.#crate_implementors_idents);
					self.0.found[#crate_implementors_indexes] = finder.find(file);
				)*

				#(
					self.0.found[#local_implementors_indexes] = self.0.finder.#local_implementors_idents.clone().find(file);
				)*
			}
		}
	};

	let impl_find = quote! {
		impl<#finder_lifetime, #generics_idents>
		#finder_wrapper_name<#finder_lifetime, #generics_idents>
		#where_clause
		{
			fn find(&mut self, file: &#finder_lifetime syn::File) -> bool{
				self.visit_file(file);
				self.0.found.iter().all(|&x| x)
			}
		}
	};

	if !already_expanded {
		struct_.attrs.push(parse_quote!(#[rust_writer_procedural::already_expanded]));
		struct_.attrs.push(parse_quote!(#[derive(Debug, Clone)]));
	}

	quote! {
		#struct_
		#impl_from_block
		#finder_wrapper
		#impl_to_find
		#impl_visit
		#impl_find
	}
}

pub(crate) fn expand_impl_finder(
	visit_lifetime: LifetimeParam,
	parsed: MacroImplParsed,
) -> TokenStream {
	let MacroImplParsed { struct_, generics_idents, where_clause } = parsed;

	let struct_name = &struct_.ident;

	let visit_lifetime = GenericParam::Lifetime(visit_lifetime);

	let visit_lifetime_gen = if generics_idents.iter().any(|generic| generic == &visit_lifetime) {
		quote! {}
	} else {
		quote! {#visit_lifetime,}
	};

	let impl_finder = quote! {
		impl<#visit_lifetime_gen #generics_idents>
		#struct_name<#generics_idents>
		#where_clause
		{
			fn find(&mut self, file: &#visit_lifetime syn::File) -> bool{
				self.visit_file(file);
				self.found.iter().all(|&x| x)
			}
		}
	};

	quote! {
		#impl_finder
	}
}
