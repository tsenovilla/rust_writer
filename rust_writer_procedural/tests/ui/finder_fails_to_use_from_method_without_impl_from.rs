// SPDX-License-Identifier: GPL-3.0

use rust_writer::ast::{
	finder::ToFind,
	implementors::{ItemToImpl, ItemToTrait},
};
use rust_writer_procedural::finder;
use syn::{parse_quote, visit::Visit, ImplItem, TraitItem};

#[finder(ItemToTrait<'a>,ItemToImpl<'a>)]
struct SomeStruct;

fn main() {
	let item_to_trait: ItemToTrait =
		("MyTrait", TraitItem::Type(parse_quote! {type Type1: From<String>;})).into();

	let item_to_impl: ItemToImpl = (
		"SomeTrait",
		"SomeImplementor",
		ImplItem::Fn(parse_quote! {
		fn some_func(&self) -> bool{
					true
				 }
		  }),
	)
		.into();

	let _some_struct: SomeStruct = (item_to_trait, item_to_impl).into();
}
