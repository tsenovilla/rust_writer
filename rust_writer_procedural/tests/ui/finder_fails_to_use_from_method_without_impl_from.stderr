error[E0277]: the trait bound `SomeStruct<'_>: From<(ItemToTrait<'_>, ItemToImpl<'_>)>` is not satisfied
  --> tests/ui/finder_fails_to_use_from_method_without_impl_from.rs:28:63
   |
28 |     let _some_struct: SomeStruct = (item_to_trait, item_to_impl).into();
   |                                                                  ^^^^ the trait `From<(ItemToTrait<'_>, ItemToImpl<'_>)>` is not implemented for `SomeStruct<'_>`
   |
   = note: required for `(ItemToTrait<'_>, ItemToImpl<'_>)` to implement `Into<SomeStruct<'_>>`
